format_doc decorator.

MSeifert04 opened the pull request:
In the discussion of #4234 @embray asked me to build a decorator to create dynamical docstrings. Since there might be other use-cases out there and I want to use it in several distinct files/classes I just thought there might be a place in the ``astropy.utils.decorators`` for it.

Please feel free to comment. If there shouldn't be use-cases outside my class or if there is already such a decorator in the python built-ins (I have looked but apparently there is no such thing) just speak up and I will only use this for my classes.

The decorator is very simple just a 4-liner but I tested it with functions, methods and it works within a python console, for sphinx and even (very happy about that) the spyder object inspector.

mhvk commented:
This looks handy indeed. Probably wise to write a test case for it...

MSeifert04 commented:
@mhvk Test cases are now also pushed. I thought I wait until I get feedback before I write tests but since that's not much trouble I just pushed them. These are very simple test cases because I'm not very familiar with ``help()`` readout comparisons so I used ``inspect.getdoc()`` (I hope this isn't considered cheating).

MSeifert04 commented:
Sorry, I forgot to test with python2. Can someone help me how to skip a test if it's python2? Should I use this command from https://pytest.org/latest/skipping.html :
```
@pytest.mark.skipif(sys.version_info < (2,8),
                    reason="requires python3")
```

MSeifert04 commented:
I'm a bit helpless with the latest test failure because I cannot reproduce the error locally. Could anyone help me?

mhvk commented:
@MSeifert04 - this looked sufficiently odd that I just restarted that test combination - maybe it is just a fluke.

MSeifert04 commented:
@mhvk Thanks for restarting the test and the error seems to be gone. I'm feeling rather stupid but I want to know: What is a fluke?

mhvk commented:
A fluke is something that happens by chance, unlikely to repeated. Of course, ideally this would not happen with deterministic machines like computers are supposed to be... But especially when networks are involved, it seems this no longer holds...

mhvk commented on astropy/utils/tests/test_decorators.py:
```diff
@@ -365,3 +365,60 @@ def _get_foo(cls): return 1
+        """this is a """
+        pass
+
+    assert inspect.getdoc(testfunc) == docstring
```
Probably over the top, but can you add one more test that shows that an existing docstring is in fact replaced?
With that, it is all OK as far as I'm concerned.

MSeifert04 replied:
I added pushed such a test, unfortunatly shortly before @embray posted his concerns.

embray commented:
I have a slightly different idea for this.  There's a simple decorator in astropy-helpers called `extends_doc`: https://github.com/astropy/astropy-helpers/blob/master/astropy_helpers/utils.py#L302

This takes the docstring of an existing class or function, and prepends its contents to the docstring of another function (or class).  In that sense it can already be used in the same way as this decorator, but it takes the docstring from some existing object, which often is how you're going to want to use it anyways.  However, I think it could also be extended to just take a string as an argument.

I don't like how the examples of this put the string in a separate variable and then pass that variable to the decorator.  You can just pass the docstring directly into the decorator (or again, inherit it from an existing object's docstring where appropriate).

I also like the idea of extending it to accept format args.  So instead of calling `.format()` explicitly, any trailing args and kwargs to the decorator can be passed in to a `.format()` call.  That way you can have a base function with a template for a docstring, but pass in arguments to the template when you use the decorator.

embray commented:
Less concerns than "Hey, you could do more with this!" :)

MSeifert04 commented:
@embray I think the use-cases might overlap but not completly. I wanted the decorator to allow that each method (or function) can be formatted. In case of taking the docstring of another method (function) with formatting placeholders like ``{0}`` will allow every other method to use the formatting but not the one from which we take it, since that would happen *before* the decorator works on the other methods. In that case I would have to create a dummy function that holds the docstring or pass the string around. Since in my case I have 4 methods that use the same docstring with only slightly different versions I would actually not mind if they pass the same string around.

The ``extend_doc`` is pretty useful but in this particular case not perfect or am I wrong about the dummy-function?

About the formatting as extra args: That's actually a pretty good idea. I tried something like this but actually using them as ``*args, **kwargs`` is pretty obvious but I missed that.

embray commented:
There are cases I've had where updating from an existing docstring is useful (regardless whether the original formatting placeholders).  But that's why I suggested allowing taking an existing object *or* a string.  I would also do away with the `replace` argument.  Instead, if you're going to allow passing in format arguments, there can also be a special format variable (say `_orig_doc_` or something like that) allowing one to (optionally) insert the content of the original docstring anywhere into the template.

MSeifert04 commented:
@mhvk Thanks about the explanation of fluke.

@embray I was wondering, I actually use a (maybe more sophisticated but in most cases unnecessary) dynamic docstring generator (https://gist.github.com/MSeifert04/28a737d6dbb24100b320) then all the questions about formatting get quite irrelevant since I can simply create the first docstring and the second one is only just a copy with ``added bits``. But that's a quite different case since I use 8 methods which share at least one mostly more parameters and the decorator is then just for getting the docstring inside the function and not a way of passing strings around.

But I thought a bit about it and created a temporary new decorator: https://gist.github.com/MSeifert04/602f6ffad653fdb05f96 . But before I push and write documentation and test cases for it I wanted to ask if that's what you were thinking of?

MSeifert04 commented:
@embray So annoying that there are no notifications on comments on ``gist``. I just saw your reply and I will update it. Should I push it into this branch or create a different one?

MSeifert04 commented:
I updated it according to @embray s suggestion.

But there may be some issues remaining (at least I'm unsure about it):
- [ ] Documentation of the decorator is quite long-ish, is it okay that way or should I move it somewhere into an main documentation ``.rst``?
- [ ] There are lots of test cases, in my opinion they cover almost everything, but maybe I missed something.
- [ ] I added the possibility to use ``'self'`` (this is a string) as docstring which is then interpreted as the functions current ``__doc__`` to cover the use-cases for ``extend_doc`` too. Is that good, bad or ``misleading``?
- [ ] Hopefully there are no test failures.
- [ ] If it is ``accepted`` should I include it in my ``NDData`` PR (because I use and need it there) or rebase the other PR?

Also if anyone happens to run across a problem in the code feel free to point me to it. :-]

MSeifert04 commented:
So seems the tests haven't found any problems (AppVeyor takes quite long, is there something wrong?). So I would be happy if anyone has any objections or comments on it, especially the points mentioned in https://github.com/astropy/astropy/pull/4242#issuecomment-148753824 .

mhvk commented:
@MSeifert04 - this is quite nice indeed!  On your points:
* How about using `'orginal_doc`' instead of `'self'`? Then, it has a consistent name between the two cases (and they would be separated by whether or not there are curly brackets. Though maybe the name should just be `__doc__` instead of `original_doc`?
* I think you should not include it in your NDData PR; easier to review things separately, and (hopefully) easy enough to rebase the NDData work.
* As is, the `astropy.utils` documentation states explicitly that these utilities are best understood through their docstring, so I think you've done the right thing (we *may* want to revisit that for decorators, but if so, I think that should be a separate issue).
* As for the name, I'd just go with `replace_docstring` or `format_docstring` -- to me at least, neither name excludes the other aspect.  Slight preference for replace.

mhvk commented on astropy/utils/decorators.py:
```diff
@@ -751,3 +752,229 @@ def _get_function_args(func, exclude_args=()):
+    Using it with ``'self'`` as docstring allows to use the decorator twice on
+    an object to first parse the new docstring and then to parse the original
+    docstring. But since this decorator performs string operations using it
+    too often might affect the performance.
```
Why might performance be an issue? This is done only when constructing the function/method, correct? I.e., it only affects setup, not runtime performance? If so, I'd remove this last sentence.

MSeifert04 replied:
You're right, it does only affect the setup time not the runtime.

MSeifert04 commented:
@mhvk 

- about ``'self'``: I wanted to seperate the case of ``using the own docstring as base`` and ``insert the own docstring in another string`` since I felt these are completly different cases. But I wasn't sure and am still not sure which name would be best. I haven't produced any code yet that was used by other people so I'm relying on feedback to produce some kind of useable API - if possible.
- same with the decorator name, the cases of replace (which kind of hides the format possibilities and the use of the own docstring as base) and format (which hides all the replace possibilities) are both not really satisfying. I originally thought of using ``extend_docstring`` (like the one in astropy-helpers but maybe having two decorators doing slightly different things might be controversial)

Sometimes I feel quite undecided about the things I do and I hope it's ok if I ask for help in here. If I should use some other kind of platform (such as the dev-mailing list) to ask such questions feel free to say so.

Thanks for your feedback! If you object to my view about those ``different cases`` I'll change the parameter names and the function name, I'm propably too close to the code myself.

embray commented:
Does this really need such a long name?

embray commented:
> How about using 'orginal_doc' instead of 'self'? Then, it has a consistent name between the two cases (and they would be separated by whether or not there are curly brackets. Though maybe the name should just be __doc__ instead of original_doc

I like `__doc__`.

MSeifert04 commented:
Ok I can live with ``'__doc__'`` and ``format_doc`` (liked that one a bit more than replace) instead of the long one. I just pushed the changes. Thanks @embray and @mhvk for reviewing.

mhvk commented:
@MSeifert04 - I think this looks great!

mhvk commented:
@MSeifert04 - looking at this again, I realised there should be an entry in `CHANGES.rst` (in the new features in 1.2 section).
@embray - otherwise, this looks ready to merge to me; do you want a last look?

MSeifert04 commented:
@mhvk Wasn't exactly sure where it would go 1.1.1 or 1.2.0. I'll add a change entry.

mhvk commented:
@MSeifert04 - I put the milestone at 1.2, since this is a new feature.

MSeifert04 commented:
@mhvk Thanks, I added the Changelog entry. Thanks (also @embray) for your patience and advice. :+1:

Let me know if any further issues arise with the code or documentation.

embray commented:
I think I have a few more minor comments on this that I'll get back to in a bit.  But looks good otherwise--thanks for your work on this.

MSeifert04 commented:
@embray I would be happy to hear about your comments.

hamogu commented:
@embray : Just doing my periodic check if we let any PRs that are almost ready to merge linger for longer than we should: Any chance you can give your few minor comments?

MSeifert04 commented:
@embray - I'd still be happy to hear about your comments.

embray commented on astropy/utils/decorators.py:
```diff
@@ -751,3 +752,230 @@ def _get_function_args(func, exclude_args=()):
+    """
+    Replaces the docstring of the decorated object and then formats it.
+
+    This decorator keeps the objects signature and only alters the saved
```
Super minor but "objects" -> "object's"

embray commented on astropy/utils/decorators.py:
```diff
@@ -751,3 +752,230 @@ def _get_function_args(func, exclude_args=()):
+    Replaces the docstring of the decorated object and then formats it.
+
+    This decorator keeps the objects signature and only alters the saved
+    ``__doc__`` property (so it cannot be used as class decorator in python2).
```
Actually, this whole sentence is confusing, now that I re-read it.  Why would this decorator be expected to change signature in the first place?  I'd just remove it.

embray commented on astropy/utils/decorators.py:
```diff
@@ -751,3 +752,230 @@ def _get_function_args(func, exclude_args=()):
+    The formatting works like :meth:`str.format` and if the decorated object
+    already has a docstring this docstring can be included in the new
+    documentation if you use the ``{__doc__}`` placeholder.
+    It's primary use is if multiple functions have the same or only
```
"It's" -> "Its"

embray commented on astropy/utils/decorators.py:
```diff
@@ -751,3 +752,230 @@ def _get_function_args(func, exclude_args=()):
+    already has a docstring this docstring can be included in the new
+    documentation if you use the ``{__doc__}`` placeholder.
+    It's primary use is if multiple functions have the same or only
+    a slightly different *long* docstring and you want to DRY
```
"It's primary use is if multiple functions have the same or only a slightly different *long* docstring ->

"Its primary use is for reusing a *long* docstring in multiple functions when it is the same or only slightly different between them."  (drop the part about DRY)

embray commented on astropy/utils/decorators.py:
```diff
@@ -751,3 +752,230 @@ def _get_function_args(func, exclude_args=()):
+        object. If it is an object like a function or class it will
+        take the docstring of this object. If it is a string it will use the
+        string itself. One special case is if the string is ``'__doc__'`` then
+        it will use the decorated functions docstring and formats it.
```
I can't decide how I feel about this "special case".  Part of me wants to just say, just use ``{{__doc__}}``.  But this is also convenient.  Maybe I'd prefer if the "special case" were just `None`.

embray commented on astropy/utils/decorators.py:
```diff
@@ -751,3 +752,230 @@ def _get_function_args(func, exclude_args=()):
+    correct docstring.
+
+    There might be problems with certain objects which have a not-writable
+    ``__doc__`` attribute. Like classes in python2.
```
This *could* be updated to support classes, even on Python 2.  I'll add something on that below.

embray commented on astropy/utils/decorators.py:
```diff
@@ -751,3 +752,230 @@ def _get_function_args(func, exclude_args=()):
+        >>> doc = '''Perform num1 + num2'''
+        >>> @format_doc(doc)
+        ... def add(num1, num2):
+        ...     return num1+num2
```
I think for the examples I'd prefer that instead of using a separate `doc` variable first, just write:
```
@format_doc('''Perform num1 + num2''')
...
```
For longer docstrings it could still be split on multiple lines.  But there's no reason a multi-line docstring can't go in the format_doc call directly.  That's a highly subjective matter of taste though.  Maybe having one example of both approaches would be fine.

embray commented on astropy/utils/decorators.py:
```diff
@@ -751,3 +752,230 @@ def _get_function_args(func, exclude_args=()):
+    on an object to first parse the new docstring and then to parse the
+    original docstring or the ``args`` and ``kwargs``.
+    """
+    def set_docstring(func):
```
Here I'd change ``func`` to ``obj`` since it could support classes.

embray commented on astropy/utils/decorators.py:
```diff
@@ -751,3 +752,230 @@ def _get_function_args(func, exclude_args=()):
+        # If the original has a not-empty docstring append it to the format
+        # kwargs.
+        kwargs['__doc__'] = func.__doc__ or ''
+        func.__doc__ = doc.format(*args, **kwargs)
```
To support classes on Python 2 we might do the following:
```python
if six.PY2 and isinstance(obj, type):
    obj = type(obj.__name__, (obj,), {'__doc__': doc.format(*args, **kwargs),
                                                         '__module__': obj.__module__})
else:
    obj.__doc__ = doc.format(*args, **kwargs)
```

It's unfortunate (and at times confusing) to insert this trivial subclass just to change the `__doc__`, but it'll get the job done for now, and probably won't be used much anyways.  But that way you can get rid of all the provisos.

embray commented:
@MSeifert04 Thanks so much for your patience on this.  This is going to be a nice little utility once it's all said and done, and I appreciate the extensive documentation and tests!

embray commented:
I just had a few other small nitpicks.  My suggestion for class docstring updates on Python 2 *should* work, but we can leave it out if we run into problems with it.

MSeifert04 commented:
@embray - Thanks for your comments. I have altered the special case when to only format the current docstring from ``"__doc__"`` to ``None``. Seemed reasonable and included your python2-class catch. I tested it and it worked for me with python 2.7. Very clever but very laborious possibility.

I did push the changes.

embray commented:
Perfect. Will merge soon as the builds succeed.

MSeifert04 commented:
One documentation (sphinx) failure because I missed a backtick.

mwcraig commented:
@MSeifert04 -- can you please rebase this? One rebase tip I don't think I've mentioned is to create a back-up branch with your current work so that if the rebase goes really badly you can start over.

MSeifert04 commented:
@mwcraig - Rebased it (without conflics apparently). I hope I haven't overwritten anything.

About the back-up-branch: Luckily this is mentioned in http://docs.astropy.org/en/stable/development/workflow/development_workflow.html#rebase. And also git provides a method to abort/undo the "last rebase" (http://stackoverflow.com/questions/134882/undoing-a-git-rebase). So I already had some back-up-plans. :)

mwcraig commented:
Merging -- thanks again, @MSeifert04!
